xdem/coreg/affine.py

```diff
@@ -648,81 +647,83 @@ def _gradient_descending_fit_func(
     """
 
     # Calculate the elevation difference
-    dh = dh_interpolator(coords_offsets[0], coords_offsets[1])
+    dh = dh_interpolator(coords_offsets[0], coords_offsets[1]).flatten()
 
-    # Return NMAD of residuals
-    return float(nmad(dh))
+    return dh
 
 
-def _gradient_descending_fit(
+def _dh_minimize_fit(
     dh_interpolator: Callable[[float, float], NDArrayf],
-    res: tuple[float, float],
-    params_noisyopt: InSpecificDict,
+    params_fit_or_bin: InFitOrBinDict,
     verbose: bool = False,
+    **kwargs: Any,
 ) -> tuple[float, float, float]:
     """
     Optimize the statistical dispersion of the elevation differences residuals.
 
     :param dh_interpolator: Interpolator returning elevation differences at the subsampled points for a certain
         horizontal offset (see _preprocess_pts_rst_subsample_interpolator).
-    :param res: Resolution of DEM.
-    :param params_noisyopt: Parameters for noisyopt minimization.
+    :param params_fit_or_bin: Parameters for fitting or binning.
     :param verbose: Whether to print statements.
 
     :return: Optimized offsets (easing, northing, vertical) in georeferenced unit.
     """
-    # Define cost function
-    def func_cost(offset: tuple[float, float]) -> float:
-        return _gradient_descending_fit_func(offset, dh_interpolator=dh_interpolator)
-
-    # Mean resolution
-    mean_res = (res[0] + res[1]) / 2
-    # Run pattern search minimization
-    res = minimizeCompass(
-        func_cost,
-        x0=tuple(x * mean_res for x in params_noisyopt["x0"]),
-        deltainit=params_noisyopt["deltainit"] * mean_res,
-        deltatol=params_noisyopt["deltatol"] * mean_res,
-        feps=params_noisyopt["feps"] * mean_res,
-        bounds=(
-            tuple(b * mean_res for b in params_noisyopt["bounds"]),
-            tuple(b * mean_res for b in params_noisyopt["bounds"]),
-        ),
-        disp=verbose,
-        errorcontrol=False,
-    )
+    # Define partial function
+    loss_func = params_fit_or_bin["fit_loss_func"]
+
+    def fit_func(coords_offsets: tuple[float, float]) -> np.floating[Any]:
+        return loss_func(_dh_minimize_fit_func(coords_offsets=coords_offsets, dh_interpolator=dh_interpolator))
+
+    # Initial offset near zero
+    init_offsets = (0, 0)
+
+    # Default parameters depending on optimizer used
+    if params_fit_or_bin["fit_minimizer"] == scipy.optimize.minimize:
+        if "method" not in kwargs.keys():
+            kwargs.update({"method": "Nelder-Mead"})
+            kwargs.update({"options": {"xatol": 10e-6, "maxiter": 500}})
+            # This method has trouble when initialized with 0,0, so defaulting to 1,1 (tip from Simon Gascoin)
```

Comment:
Perhaps give a permalink to the PR comment

Reply:
Done!